File: src/merge_sort/merge_sort.rb
require 'rspec'
require 'pry'

def merge_sort(array)
  return array if array.length <= 1
  slice_size = array.length / 2
  left = array[0...slice_size]
  right = array[slice_size...array.length]
  merge(array, merge_sort(left), merge_sort(right))
end

def merge(array, left, right)
  i = 0
  j = 0
  result = []
  (0...array.length).each do |k|
    if !left[i].nil? && (right[j].nil? || left[i] < right[j])
      result[k] = left[i]
      i += 1
    else
      result[k] = right[j]
      j += 1
    end
  end
  result
end

describe 'merge_sort' do
  let(:input) { [3, 2, 5, 4, 1] }

  subject { merge_sort(input) }

  it { is_expected.to eq [1, 2, 3, 4, 5] }

  context 'empty' do
    let(:input) { [] }
    it { is_expected.to eq [] }
  end

  context '1 eement' do
    let(:input) { [5] }

    it { is_expected.to eq [5] }
  end
end

File: src/merge_sort/merge_sort.exs
defmodule MergeSort do
  def sort([]), do: []
  def sort(arr = [_]), do: arr
  def sort(arr) do
    {left, right} = split(arr)
    merge(sort(left), sort(right))
  end

  def merge(left, right), do: merge(left, right, [])
  def merge([], [], result), do: Enum.reverse(result)
  def merge([head | tail], [], result), do: merge(tail, [], [head | result])
  def merge([], [head | tail], result), do: merge([], tail, [head | result])

  def merge([lhead | ltail], [rhead | rtail], result) when lhead < rhead do
    merge(ltail, [rhead | rtail], [lhead | result])
  end

  def merge([lhead | ltail], [rhead | rtail], result) when lhead >= rhead do
    merge([lhead | ltail], rtail, [rhead | result])
  end

  def split(arr) do
    mid = length(arr) / 2 |> round
    arr |> Enum.split(mid)
  end
end

ExUnit.start
defmodule MergeSortTest do
  use ExUnit.Case

  setup_all do
    { :ok, input: [3, 4, 2, 5, 2, 1] }
  end

  test "sorting empty" do
    assert MergeSort.sort([]) == []
  end

  test "sorting works", state do
    assert MergeSort.sort(state[:input]) == [1, 2, 2, 3, 4, 5]
  end
end
